rove_connector/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+    #[postgres(name = "hag")]
+    _hag: Option<f32>,
+}
+
+fn extract_time_spec(
+    time_spec: &TimeSpec,
+    num_leading_points: u8,
+    num_trailing_points: u8,
+) -> Result<(DateTime<Utc>, DateTime<Utc>), data_switch::Error> {
+    // TODO: should time_spec just use chrono timestamps instead of unix?
+    // IIRC the reason for unix timestamps was easy compatibility with protobuf, but that's
+    // less of a priority now
+    let start_time = Utc.timestamp_opt(time_spec.timerange.start.0, 0).unwrap()
+        - (time_spec.time_resolution * num_leading_points.into());
+    // TODO: figure out whether the range in postgres is inclusive on the range here or
+    // we need to add 1 second
+    let end_time = Utc.timestamp_opt(time_spec.timerange.end.0, 0).unwrap()
+        + (time_spec.time_resolution * num_trailing_points.into());
+
+    Ok((start_time, end_time))
+}
+
+// TODO: does the input type match postgres-types?
+fn regularize(
+    obses: Vec<Obs>,
+    start_time: DateTime<Utc>,
+    end_time: DateTime<Utc>,
+    time_resolution: RelativeDuration,
+    expected_len: usize,
+) -> Vec<Option<f32>> {
+    let mut out = Vec::with_capacity(expected_len);
+    let mut curr_obs_time = start_time;
+
+    for obs in obses {
+        while curr_obs_time < obs.obstime {
+            out.push(None);
+            curr_obs_time = curr_obs_time + time_resolution;
+        }
+        if curr_obs_time == obs.obstime {
+            out.push(Some(obs.obsvalue));
+            curr_obs_time = curr_obs_time + time_resolution;
+        } else {
+            // In this case the observation is misaligned, so we should skip it. There's a case
+            // to be made for returning an error, but I think we ought to be more robust.
+            continue;
+        }
+    }
+
+    while curr_obs_time <= end_time {
+        out.push(None);
+        curr_obs_time = curr_obs_time + time_resolution;
+    }
+
+    out
+}
+
+impl Connector {
+    async fn fetch_one(
+        &self,
+        ts_id: i32,
+        time_spec: &TimeSpec,
+        num_leading_points: u8,
+        num_trailing_points: u8,
+    ) -> Result<DataCache, data_switch::Error> {
+        // TODO: matching intervals like this is a hack, but currently necessary to avoid
+        // SQL injection. Ideally we could pass an interval type as a query param, which would
+        // also save us the query_string allocation, but no ToSql implementations for intervals
+        // currently exist in tokio_postgres, so we need to implement it ourselves.
+        let interval = match time_spec.time_resolution {
+            x if x == RelativeDuration::minutes(1) => "1 minute",
+            x if x == RelativeDuration::hours(1) => "1 hour",
+            x if x == RelativeDuration::days(1) => "1 day",
+            _ => {
+                return Err(data_switch::Error::Other(Box::new(
+                    Error::UnhandledTimeResolution(time_spec.time_resolution),
+                )))
+            }
+        };
```

Comment:
The problem with using an enum, is that we should be forward-compatible with time_resolutions we don't expect, and RelativeDuration is maximally flexible that way (it's also nice that it implements math operations with `chrono::DateTime`). I think this might be more neatly solvable by implementing `ToSql` for `RelativeDuration`.

More importantly, this code is only relevant to the SpaceSpec::One case, and as I note in another `TODO`, we might find out in benchmarking that this approach is less efficient, and then just reimplement this as a degenerate case of the SpaceSpec::All handling, in which case we can get rid of this entirely.

Reply:
But I don't think we can implement `ToSql` for `RelativeDuration` here, it needs to be done in `chrono` directly. 
Anyway I agree it'd be nice if both fetch functions were implemented in the same way

Reply:
> But I don't think we can implement ToSql for RelativeDuration here, it needs to be done in chrono directly.

I've contributed upstream in chronoutil before, there's a good chance the maintainer will be amenable to this. If not we can always vendor it.

> Anyway I agree it'd be nice if both fetch functions were implemented in the same way

And I don't think it's possible to implement `fetch_all` this way, because I couldn't find a way to `JOIN` inside a `GROUP BY` aggregation since the row set is implicit. That alone may be a reason to ditch this implementation.

Reply:
No, I meant using the same strategy we have in `fetch_all` here (the same thing you wanted to do as I understand it)

Reply:
Agree, but I think we should wait until we benchmark it and for now just leave the TODO